Fixing for Bitbucket 8.9

voarsh2 opened the pull request:
I had to revert this commit: https://github.com/caelumlux/atlassian-agent/commit/1c05244135a38a3b088c4f3913b5c3ce7936555b

I was running on Bitbucket 8.9.2 and licence verification wasn't working. After reverting, and building the agent it was working: As seen in this issue: https://github.com/haxqer/bitbucket/issues/1#issuecomment-1620650035. Using DC Helm Chart:
```
additionalJvmArgs:
    - '-javaagent:/var/agent/atlassian-agent-jar-with-dependencies.jar'
```

Dockerfile:
```
FROM atlassian/bitbucket:8.9.2

LABEL maintainer="haxqer <haxqer666@gmail.com>" version="8.9.2"

ARG BITBUCKET_VERSION=8.9.2
ARG BITBUCKET_PRODUCT=bitbucket-software
ARG AGENT_VERSION=1.3.3
ARG MYSQL_DRIVER_VERSION=8.0.22

ENV BITBUCKET_USER=bitbucket \
    BITBUCKET_GROUP=bitbucket \
    BITBUCKET_HOME=/var/atlassian/application-data/bitbucket \
    BITBUCKET_INSTALL=/opt/atlassian/bitbucket \
    JVM_MINIMUM_MEMORY=2g \
    ATLASSIAN_DIR=/opt/atlassian/bitbucket/app/WEB-INF/lib/WEB-INF/lib \
    JVM_MAXIMUM_MEMORY=6g \
    JVM_CODE_CACHE_ARGS='-XX:InitialCodeCacheSize=2g -XX:ReservedCodeCacheSize=6g' \
    AGENT_PATH=/var/agent \
    AGENT_FILENAME=atlassian-agent-jar-with-dependencies.jar

ENV JAVA_OPTS="-javaagent:${AGENT_PATH}/${AGENT_FILENAME} ${JAVA_OPTS}"

COPY ./atlassian-agent-jar-with-dependencies.jar ${AGENT_PATH}/${AGENT_FILENAME}

#RUN mkdir -p ${BITBUCKET_INSTALL} ${BITBUCKET_HOME} ${AGENT_PATH} \
# && 
#RUN mkdir ${AGENT_PATH}
#RUN curl -o ${AGENT_PATH}/${AGENT_FILENAME} https://github.com/haxqer/jira/releases/download/v1.3.3/atlassian-agent.jar -L
#&& curl -o /tmp/atlassian.tar.gz https://product-downloads.atlassian.com/software/stash/downloads/atlassian-bitbucket-${BITBUCKET_VERSION}.tar.gz -L \
#&& tar xzf /tmp/atlassian.tar.gz -C ${BITBUCKET_INSTALL}/ --strip-components 1 \
#&& rm -f /tmp/atlassian.tar.gz \
#&& curl -o ${BITBUCKET_INSTALL}/app/WEB-INF/lib/mysql-connector-java-${MYSQL_DRIVER_VERSION}.jar https://repo1.maven.org/maven2/mysql/mysql-connector-java/${MYSQL_DRIVER_VERSION}/mysql-connector-java-${MYSQL_DRIVER_VERSION}.jar -L

RUN export ATLASSIAN_DIR
RUN export JAVA_OPTS="-javaagent:/var/agent/atlassian-agent-jar-with-dependencies.jar ${JAVA_OPTS}"

#RUN export CONTAINER_USER=$BITBUCKET_USER \
#&& export CONTAINER_GROUP=$BITBUCKET_GROUP \
#&& groupadd -r $BITBUCKET_GROUP && useradd -r -g $BITBUCKET_GROUP $BITBUCKET_USER \
#&& chown -R $BITBUCKET_USER:$BITBUCKET_GROUP ${BITBUCKET_INSTALL} ${BITBUCKET_HOME}/ ${AGENT_PATH}

#VOLUME $BITBUCKET_HOME
#USER $BITBUCKET_USER
#WORKDIR $BITBUCKET_INSTALL
#EXPOSE 7990
#EXPOSE 7999

#ENTRYPOINT ["/opt/atlassian/bitbucket/bin/start-bitbucket.sh", "-fg"]
```


However, plugins don't seem to pass licence verification now - it assumes a "standard" licence.

@caelumlux are you able to offer some help into this?

voarsh committed:
Update src/main/java/io/zhile/crack/atlassian/agent/KeyTransformer.java

voarsh committed:
Revert "Update src/main/java/io/zhile/crack/atlassian/license/LicenseProperty.java"

This reverts commit 32afddfecfef65d1d1ae9b09f98ae03c93ec494f.

voarsh committed:
Update src/main/java/io/zhile/crack/atlassian/license/LicenseProperty.java

voarsh committed:
Revert "Update src/main/java/io/zhile/crack/atlassian/license/products/ThirdPlugin.java"

This reverts commit b10798f1655a73948bb999055e20ede2c333e063.

voarsh committed:
Update src/main/java/io/zhile/crack/atlassian/license/products/ThirdPlugin.java

voarsh committed:
Update src/main/java/io/zhile/crack/atlassian/license/products/ThirdPlugin.java

voarsh committed:
Merge remote-tracking branch 'bitbucket/master'

voarsh committed:
Remove DataCenter and Subscription properties when not applicable

- In LicenseProperty.java, the code was modified to remove the DataCenter and Subscription properties when they are not applicable.
- In ThirdPlugin.java, the constructor was modified to remove the dataCenter parameter since it is no longer needed.

caelumlux commented:
@voarsh2 

I'm not very familiar with Dockerfiles.

I think your commit is unnecessary I saw that you have already set the environment variable in your docker file，like this :

`ATLASSIAN_DIR=/opt/atlassian/bitbucket/app/WEB-INF/lib/WEB-INF/lib`


The license for plugins should be generated like this :
`java -jar atlassian-agent.jar -p com.gliffy.integration.confluence -m aaa@bbb.com -n my_name -o https://zhile.io -s ABCD-1234-EFGH-5678`


I see that the command you used to generate the license is like this. 
`java -jar /var/agent/atlassian-agent-jar-with-dependencies.jar -p bitbucket  -m haxqer666@gmail.com     -n haxqer666@gmail.com     -o http://website     -s xxxxx -d`

Maybe, you shoud change "-p bitbucket" to "-p [plugin id]", the plugin id is like package name, such as "com.gliffy.integration.confluence".

voarsh2 commented:
@caelumlux 

> Maybe, you shoud change "-p bitbucket" to "-p [plugin id]", the plugin id is like package name, such as "com.gliffy.integration.confluence".

Sorry, perhaps you misread somewhere, or my output I pasted the wrong thing. BUT:

> I think your commit is unnecessary I saw that you have already set the environment variable in your docker file，like this 
ATLASSIAN_DIR=/opt/atlassian/bitbucket/app/WEB-INF/lib/WEB-INF/lib

Using your original code, having the ```ATLASSIAN_DIR``` didn't do anything to create a "valid" licence when trying to generate a DC Bitbucket licence, when I reverted, it created a "valid" Bitbucket DC licence.

The commit I reverted actually let me have a valid DC licence FOR Bitbucket (8.9).  It wasn't working without that commit being reverted.

For the plugin, I didn't do ```-p bitbucket``` - but it set the plugin licence as standard, despite specifying the -d flag....

To which, I say, there's something different for plugins in Bitbucket 8.9, making the current agent not work.

Here's some screenshots and info:

Without reverting the commit, with an existing licence generated (with the reverted commit), Bitbucket would complain the licence is invalid. When I revert the commit I mentioned, it works:
```
2023-07-05 15:43:12,552 INFO [spring-startup] c.a.s.internal.home.HomeLockAcquirer Successfully acquired lock on home directory /var/atlassian/application-data/bitbucket
Wed, Jul 5 2023 4:43:30 pm | atlassian-agent: skip verify the license.
Wed, Jul 5 2023 4:43:30 pm | atlassian-agent: skip license hash check
Wed, Jul 5 2023 4:43:30 pm | 2023-07-05 15:43:30,558 INFO [spring-startup] c.a.s.internal.home.HomeLockAcquirer Successfully acquired lock on home directory /var/atlassian/application-data/shared-home
Wed, Jul 5 2023 4:43:31 pm | WARNING: An illegal reflective access operation has occurred
Wed, Jul 5 2023 4:43:31 pm | WARNING: Illegal reflective access by com.hazelcast.internal.networking.nio.SelectorOptimizer (file:/opt/atlassian/bitbucket/app/WEB-INF/lib/hazelcast-3.12.13.jar) to field sun.nio.ch.SelectorImpl.selectedKeys
Wed, Jul 5 2023 4:43:31 pm | WARNING: Please consider reporting this to the maintainers of com.hazelcast.internal.networking.nio.SelectorOptimizer
Wed, Jul 5 2023 4:43:31 pm | WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Wed, Jul 5 2023 4:43:31 pm | WARNING: All illegal access operations will be denied in a future release
Wed, Jul 5 2023 4:43:31 pm | 2023-07-05 15:43:31,896 WARN [spring-startup] c.h.s.d.integration.DiscoveryService [10.42.42.110]:5701 [bittesting-bitbucket] [3.12.13] Cannot fetch the current zone, ZONE_AWARE feature is disabled
Wed, Jul 5 2023 4:43:31 pm | 2023-07-05 15:43:31,943 WARN [spring-startup] c.h.kubernetes.KubernetesClient Cannot fetch public IPs of Hazelcast Member PODs, you won't be able to use Hazelcast Smart Client from outside of the Kubernetes network
Wed, Jul 5 2023 4:44:15 pm | 2023-07-05 15:44:15,443 INFO [spring-startup] c.a.s.i.s.g.m.DefaultSidecarManager Sidecar started after 10051ms
```

## Generating a Bitbucket DC licence:
When https://github.com/caelumlux/atlassian-agent/commit/1c05244135a38a3b088c4f3913b5c3ce7936555b is reverted:
```
java -jar /var/agent/atlassian-agent-jar-with-dependencies.jar -p bitbucket  -m haxqer666@gmail.com     -n haxqer666@gmail.com     -o http://website     -s xxxxx -d
```
![image](https://github.com/caelumlux/atlassian-agent/assets/60437695/cc928164-a6c5-4c69-88a4-ecd492d215e1)



## Plugin is assigned standard licence, incompatible with DC Bitbucket licence:
(with https://github.com/caelumlux/atlassian-agent/commit/1c05244135a38a3b088c4f3913b5c3ce7936555b reverted)
```
java -jar /var/agent/atlassian-agent-jar-with-dependencies.jar -p com.simplenia.stash.plugins.snippets  -m haxqer666@gmail.com     -n haxqer666@gmail.com     -o http://website     -s xxxxx -d
```

![image](https://github.com/caelumlux/atlassian-agent/assets/60437695/ca27744f-fbce-451a-a23e-bb7f8758e271)


So, as you see, things aren't quite working right.... Have you confirmed Bitbucket 8.9 is working with your modifications?



# Use original, (your version) without modification:

## Start up Bitbucket with a licence from the agent:

```
lib路径为：/opt/atlassian/bitbucket/app/WEB-INF/lib/WEB-INF/lib
--
Wed, Jul 5 2023 5:06:37 pm | java.lang.NullPointerException
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.util.Objects.requireNonNull(Objects.java:221)
Wed, Jul 5 2023 5:06:37 pm | at io.zhile.crack.atlassian.agent.KeyTransformer.handleLicenseDecoder(KeyTransformer.java:88)
Wed, Jul 5 2023 5:06:37 pm | at io.zhile.crack.atlassian.agent.KeyTransformer.transform(KeyTransformer.java:33)
Wed, Jul 5 2023 5:06:37 pm | at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:246)
Wed, Jul 5 2023 5:06:37 pm | at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
Wed, Jul 5 2023 5:06:37 pm | at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:563)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.ClassLoader.defineClass1(Native Method)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:555)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:458)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:452)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.security.AccessController.doPrivileged(Native Method)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:451)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:135)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.Class.forName0(Native Method)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.Class.forName(Class.java:398)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.boot.web.embedded.tomcat.TomcatEmbeddedWebappClassLoader.loadFromParent(TomcatEmbeddedWebappClassLoader.java:114)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.boot.web.embedded.tomcat.TomcatEmbeddedWebappClassLoader.doLoadClass(TomcatEmbeddedWebappClassLoader.java:87)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.boot.web.embedded.tomcat.TomcatEmbeddedWebappClassLoader.loadClass(TomcatEmbeddedWebappClassLoader.java:70)
Wed, Jul 5 2023 5:06:37 pm | at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1220)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.Class.forName0(Native Method)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.Class.forName(Class.java:398)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.util.ClassUtils.forName(ClassUtils.java:284)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.AbstractBeanDefinition.resolveBeanClass(AbstractBeanDefinition.java:469)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.AbstractBeanFactory.doResolveBeanClass(AbstractBeanFactory.java:1621)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.AbstractBeanFactory.resolveBeanClass(AbstractBeanFactory.java:1548)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:704)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:674)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1684)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:570)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:542)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:103)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:748)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:564)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.web.servlet.FrameworkServlet.configureAndRefreshWebApplicationContext(FrameworkServlet.java:702)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.web.servlet.FrameworkServlet.createWebApplicationContext(FrameworkServlet.java:668)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.web.servlet.FrameworkServlet.createWebApplicationContext(FrameworkServlet.java:716)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.web.servlet.FrameworkServlet.initWebApplicationContext(FrameworkServlet.java:591)
Wed, Jul 5 2023 5:06:37 pm | at com.atlassian.johnson.spring.web.servlet.JohnsonDispatcherServlet.initWebApplicationContext(JohnsonDispatcherServlet.java:191)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.web.servlet.FrameworkServlet.initServletBean(FrameworkServlet.java:530)
Wed, Jul 5 2023 5:06:37 pm | at org.springframework.web.servlet.HttpServletBean.init(HttpServletBean.java:170)
Wed, Jul 5 2023 5:06:37 pm | at javax.servlet.GenericServlet.init(GenericServlet.java:158)
Wed, Jul 5 2023 5:06:37 pm | at com.atlassian.johnson.spring.lifecycle.LifecycleDispatcherServlet.lambda$init$0(LifecycleDispatcherServlet.java:87)
Wed, Jul 5 2023 5:06:37 pm | at java.base/java.lang.Thread.run(Thread.java:829)
Wed, Jul 5 2023 5:06:39 pm | 2023-07-05 16:06:39,941 INFO [spring-startup] c.a.s.internal.home.HomeLockAcquirer Successfully acquired lock on home directory /var/atlassian/application-data/bitbucket
Wed, Jul 5 2023 5:07:03 pm | ============================== agent working ==============================
Wed, Jul 5 2023 5:07:03 pm | 2023-07-05 16:07:03,693 WARN [spring-startup] c.a.s.i.license.DefaultLicenseHelper The provided license could not be verified
Wed, Jul 5 2023 5:07:03 pm | com.atlassian.extras.decoder.api.LicenseVerificationException: Failed to verify license hash: missing property licenseHash
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.extras.decoder.v2.Version2LicenseDecoder.verifyLicenseHash(Version2LicenseDecoder.java:152)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.extras.decoder.v2.Version2LicenseDecoder.doDecode(Version2LicenseDecoder.java:119)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.extras.decoder.api.AbstractLicenseDecoder.decode(AbstractLicenseDecoder.java:15)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.extras.core.DefaultLicenseManager.decodeLicense(DefaultLicenseManager.java:62)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.extras.core.DefaultLicenseManager.getAtlassianLicense(DefaultLicenseManager.java:50)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.extras.core.DefaultLicenseManager.getLicense(DefaultLicenseManager.java:39)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.decode(DefaultLicenseHelper.java:67)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.getDecoded(DefaultLicenseHelper.java:108)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.isClusteringEnabled(DefaultLicenseHelper.java:117)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.stash.internal.home.SharedHomeLockAcquirer.isSharedLockRequired(SharedHomeLockAcquirer.java:106)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.stash.internal.home.SharedHomeLockAcquirer.createHomeLock(SharedHomeLockAcquirer.java:96)
Wed, Jul 5 2023 5:07:03 pm | at com.atlassian.stash.internal.home.HomeLockAcquirer.lock(HomeLockAcquirer.java:56)
Wed, Jul 5 2023 5:07:03 pm | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Wed, Jul 5 2023 5:07:03 pm | at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Wed, Jul 5 2023 5:07:03 pm | at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Wed, Jul 5 2023 5:07:03 pm | at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:920)
Wed, Jul 5 2023 5:07:03 pm | at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583)
Wed, Jul 5 2023 5:07:03 pm | at javax.servlet.GenericServlet.init(GenericServlet.java:158)
Wed, Jul 5 2023 5:07:03 pm | at java.base/java.lang.Thread.run(Thread.java:829)
Wed, Jul 5 2023 5:07:03 pm | ... 37 frames trimmed
Wed, Jul 5 2023 5:07:03 pm | 2023-07-05 16:07:03,731 INFO [spring-startup] c.a.s.internal.home.HomeLockAcquirer Successfully acquired lock on home directory /var/atlassian/application-data/shared-home
Wed, Jul 5 2023 5:07:07 pm | WARNING: An illegal reflective access operation has occurred
Wed, Jul 5 2023 5:07:07 pm | WARNING: Illegal reflective access by com.hazelcast.internal.networking.nio.SelectorOptimizer (file:/opt/atlassian/bitbucket/app/WEB-INF/lib/hazelcast-3.12.13.jar) to field sun.nio.ch.SelectorImpl.selectedKeys
Wed, Jul 5 2023 5:07:07 pm | WARNING: Please consider reporting this to the maintainers of com.hazelcast.internal.networking.nio.SelectorOptimizer
Wed, Jul 5 2023 5:07:07 pm | WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Wed, Jul 5 2023 5:07:07 pm | WARNING: All illegal access operations will be denied in a future release
Wed, Jul 5 2023 5:07:07 pm | 2023-07-05 16:07:07,460 WARN [spring-startup] c.h.s.d.integration.DiscoveryService [10.42.189.114]:5701 [bittesting-bitbucket] [3.12.13] Cannot fetch the current zone, ZONE_AWARE feature is disabled
Wed, Jul 5 2023 5:07:07 pm | 2023-07-05 16:07:07,606 WARN [spring-startup] c.h.kubernetes.KubernetesClient Cannot fetch public IPs of Hazelcast Member PODs, you won't be able to use Hazelcast Smart Client from outside of the Kubernetes network
Wed, Jul 5 2023 5:07:07 pm | 2023-07-05 16:07:07,665 WARN [hz.hazelcast.cached.thread-3] c.a.s.i.license.DefaultLicenseHelper The provided license could not be verified
Wed, Jul 5 2023 5:07:07 pm | com.atlassian.extras.decoder.api.LicenseVerificationException: Failed to verify license hash: missing property licenseHash
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.extras.decoder.v2.Version2LicenseDecoder.verifyLicenseHash(Version2LicenseDecoder.java:152)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.extras.decoder.v2.Version2LicenseDecoder.doDecode(Version2LicenseDecoder.java:119)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.extras.decoder.api.AbstractLicenseDecoder.decode(AbstractLicenseDecoder.java:15)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.extras.core.DefaultLicenseManager.decodeLicense(DefaultLicenseManager.java:62)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.extras.core.DefaultLicenseManager.getAtlassianLicense(DefaultLicenseManager.java:50)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.extras.core.DefaultLicenseManager.getLicense(DefaultLicenseManager.java:39)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.decode(DefaultLicenseHelper.java:67)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.getDecoded(DefaultLicenseHelper.java:108)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.isClusteringEnabled(DefaultLicenseHelper.java:117)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.stash.internal.cluster.DefaultClusterJoinManager.checkAcceptingClusterConnections(DefaultClusterJoinManager.java:241)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.stash.internal.cluster.DefaultClusterJoinManager.connect(DefaultClusterJoinManager.java:139)
Wed, Jul 5 2023 5:07:07 pm | at com.atlassian.stash.internal.hazelcast.ClusterJoinSocketInterceptor.onConnect(ClusterJoinSocketInterceptor.java:56)
Wed, Jul 5 2023 5:07:07 pm | at com.hazelcast.nio.NodeIOService.interceptSocket(NodeIOService.java:302)
Wed, Jul 5 2023 5:07:07 pm | at com.hazelcast.nio.tcp.TcpIpConnector$ConnectTask.tryToConnect(TcpIpConnector.java:185)
Wed, Jul 5 2023 5:07:07 pm | at com.hazelcast.nio.tcp.TcpIpConnector$ConnectTask.run(TcpIpConnector.java:115)
Wed, Jul 5 2023 5:07:07 pm | at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:227)
Wed, Jul 5 2023 5:07:07 pm | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
Wed, Jul 5 2023 5:07:07 pm | at java.base/java.lang.Thread.run(Thread.java:829)
Wed, Jul 5 2023 5:07:07 pm | at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
Wed, Jul 5 2023 5:07:07 pm | at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Wed, Jul 5 2023 5:07:07 pm | ... 1 frame trimmed
Wed, Jul 5 2023 5:07:11 pm | 2023-07-05 16:07:11,872 WARN [hz.hazelcast.cached.thread-3] c.a.s.i.license.DefaultLicenseHelper The provided license could not be verified
Wed, Jul 5 2023 5:07:11 pm | com.atlassian.extras.decoder.api.LicenseVerificationException: Failed to verify license hash: missing property licenseHash
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.extras.decoder.v2.Version2LicenseDecoder.verifyLicenseHash(Version2LicenseDecoder.java:152)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.extras.decoder.v2.Version2LicenseDecoder.doDecode(Version2LicenseDecoder.java:119)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.extras.decoder.api.AbstractLicenseDecoder.decode(AbstractLicenseDecoder.java:15)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.extras.core.DefaultLicenseManager.decodeLicense(DefaultLicenseManager.java:62)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.extras.core.DefaultLicenseManager.getAtlassianLicense(DefaultLicenseManager.java:50)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.extras.core.DefaultLicenseManager.getLicense(DefaultLicenseManager.java:39)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.decode(DefaultLicenseHelper.java:67)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.getDecoded(DefaultLicenseHelper.java:108)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.license.DefaultLicenseHelper.isClusteringEnabled(DefaultLicenseHelper.java:117)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.cluster.LicenseClusterJoinCheck.isClusteringEnabledResult(LicenseClusterJoinCheck.java:61)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.cluster.LicenseClusterJoinCheck.connect(LicenseClusterJoinCheck.java:40)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.cluster.DefaultClusterJoinManager.connect(DefaultClusterJoinManager.java:165)
Wed, Jul 5 2023 5:07:11 pm | at com.atlassian.stash.internal.hazelcast.ClusterJoinSocketInterceptor.onConnect(ClusterJoinSocketInterceptor.java:56)
Wed, Jul 5 2023 5:07:11 pm | at com.hazelcast.nio.NodeIOService.interceptSocket(NodeIOService.java:302)
Wed, Jul 5 2023 5:07:11 pm | at com.hazelcast.nio.tcp.TcpIpConnector$ConnectTask.tryToConnect(TcpIpConnector.java:185)
Wed, Jul 5 2023 5:07:11 pm | at com.hazelcast.nio.tcp.TcpIpConnector$ConnectTask.run(TcpIpConnector.java:115)
Wed, Jul 5 2023 5:07:11 pm | at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:227)
Wed, Jul 5 2023 5:07:11 pm | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
Wed, Jul 5 2023 5:07:11 pm | at java.base/java.lang.Thread.run(Thread.java:829)
Wed, Jul 5 2023 5:07:11 pm | at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
Wed, Jul 5 2023 5:07:11 pm | at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Wed, Jul 5 2023 5:07:11 pm | ... 1 frame trimmed
```

## Plugins without reverting the commit cannot be licenced at all (doesn't even get a "server" licence, which is the wrong type, anyway.

voarsh2 commented:
@caelumlux Not sure if you got pinged about my previous comment

caelumlux commented:
@voarsh2 
Sorry, i find the Env is wrong in the readme,
If the product is bitbucket, the ENV shoud be `ATLASSIAN_DIR="/opt/atlassian/bitbucket/app/WEB-INF/lib"`.

if you set the ENV correctly,the log will be the next, when you run image.
```
2023-07-06 09:00:35,490 INFO  [threadpool:thread-2] @8CCSJBx540x17x0 1ky2f7b 172.17.0.1 "POST /setup HTTP/1.1" c.a.s.i.b.l.DefaultLiquibaseMigrationDao Connecting to target database
2023-07-06 09:00:35,949 INFO  [threadpool:thread-2] @8CCSJBx540x17x0 1ky2f7b 172.17.0.1 "POST /setup HTTP/1.1" c.a.s.i.b.l.DefaultLiquibaseMigrationDao Restore process completed
atlassian-agent: skip verify the license.
atlassian-agent: skip license hash check
2023-07-06 09:02:07,039 WARN  [http-nio-7990-exec-9] @8CCSJBx542x21x0 1ky2f7b 172.17.0.1 "POST /setup HTTP/1.1" c.a.s.i.license.DefaultLicenseHelper The decoded license does not include a Bitbucket license. Products licensed: []
atlassian-agent: skip verify the license.
atlassian-agent: skip license hash check
2023-07-06 09:04:06,291 INFO  [AtlassianEvent::thread-2] @8CCSJBx544x25x0 1ky2f7b 172.17.0.1 "POST /setup HTTP/1.1" c.a.b.internal.ssh.server.SshServer Starting SSH server on port 7999...
2023-07-06 09:04:06,392 INFO  [AtlassianEvent::thread-2] @8CCSJBx544x25x0 1ky2f7b 172.17.0.1 "POST /setup HTTP/1.1" c.a.b.internal.ssh.server.SshServer Started SSH server successfully.
2023-07-06 09:04:06,448 INFO  [AtlassianEvent::thread-2] @8CCSJBx544x25x0 1ky2f7b 172.17.0.1 "POST /setup HTTP/1.1" c.a.b.i.s.c.j.c.HealthCheckRunner New health check registered: AuthenticationCheck
```

but I don't dou why you failure to crack plugin, I'm success.

generate lisence commond `java -jar atlassian-agent-jar-with-dependencies.jar -p com.simplenia.stash.plugins.snippets -m example@example.com -n example.com-o http://example.com-s XXXX-XXXX-XXXX-XXXX`

the log is :
```
2023-07-06 09:06:18,398 WARN  [http-nio-7990-exec-7 url: /rest/plugins/1.0/installed-marketplace; user: admin] admin @8CCSJBx546x159x1 1ky2f7b 172.17.0.1 "GET /rest/plugins/1.0/installed-marketplace HTTP/1.1" com.atlassian.upm.pac.PacClientImpl The request to check for app updates may take longer than expected because 95 apps are installed that could have updates. This may impact the performance of loading the Manage Apps page.
atlassian-agent: skip verify the license.
atlassian-agent: skip verify the license.
atlassian-agent: skip license hash check
atlassian-agent: skip verify the license.
atlassian-agent: skip license hash check
atlassian-agent: skip verify the license.
atlassian-agent: skip license hash check
```

and this is the result:
![image](https://github.com/caelumlux/atlassian-agent/assets/54264777/4373e651-43e8-4502-96f6-556ca01190d3)

i will push the docker image,and pull read me. you can wait for a little.

the docker file is :
```
FROM atlassian/bitbucket:8.11.2
LABEL maintainer="caelumlux <caelumlux@outlook.com>" version="8.11.2"

ARG AGENT_PATH=/opt/atlassian
ARG AGENT_FILENAME=atlassian-agent-jar-with-dependencies.jar
# 将破解包加入容器
COPY ./atlassian-agent-jar-with-dependencies.jar /opt/atlassian/
ENV ATLASSIAN_DIR="/opt/atlassian/bitbucket/app/WEB-INF/lib"
ENV JAVA_OPTS="-javaagent:${AGENT_PATH}/${AGENT_FILENAME} ${JAVA_OPTS}"
```

docker-comose.yml is:

```
version: "3"

services:
  confluence:
    image: caelumlux/bitbucket:8.11.2
    container_name: bitbucket
    ports:
      - "7990:7990"
      - "7999:7999"
    volumes:
      - "/var/atlassian/application-data/bitbucket:/var/atlassian/application-data/bitbucket "
    extra_hosts:
      - "host.docker.internal:host-gateway"
    environment:
      - TZ="Asia/Shanghai"
    network_mode: bridge
```

caelumlux commented:
@voarsh2  maybe you should remove environment in the docker-comose.yml

voarsh2 commented:
I will need to re-read your last reply. But, I wanted to quickly highlight THIS:


> but I don't dou why you failure to crack plugin, I'm success.
> 
> generate lisence commond `java -jar atlassian-agent-jar-with-dependencies.jar -p com.simplenia.stash.plugins.snippets -m example@example.com -n example.com-o http://example.com-s XXXX-XXXX-XXXX-XXXX`
> 
> the log is :

This is a STANDARD licence/server licence, not a Data Center licence.............
It can create a DC licence for the install, but PLUGIN is only STANDARD, even if you want a DC one...
When I have a DC licence for install, I can't use server plugins...........

Try again with "-d" flag for data center. (DC licence for install, DC for plugin, you'll see it won't work)

caelumlux commented:
Maybe the agent doesn't work in the DC . You can fix it ....

voarsh2 commented:
@caelumlux please read this important comment regarding the fact that you generated a server install licence, with a server type plugin licence... DC/Data Center licence for installation works, but not for plugin: https://github.com/caelumlux/atlassian-agent/pull/1#issuecomment-1623319535 - try with ```-d``` flag to try and make a data center install licence and data center plugin licence


> Sorry, i find the Env is wrong in the readme, If the product is bitbucket, the ENV shoud be `ATLASSIAN_DIR="/opt/atlassian/bitbucket/app/WEB-INF/lib"`.
> 
> if you set the ENV correctly,the log will be the next, when you run image.

Yes, I thought it looked wrong, which is why I had tried to specify it as "/opt/atlassian/bitbucket/app/WEB-INF/lib" in some other attempts. But worked when I revert your commit regarding to location of the libraries........
Will look to try again (I've been trying to sort this for hours and hours. tiring...) and make sure ENV "ATLASSIAN_DIR is not specified......

I will try 8.11 as well, however, there could also be changed in 8.9 different from 8.11. I am using 8.9 because it is long term supported.



> i will push the docker image,and pull read me. you can wait for a little.
> 
> the docker file is :
> 
> ```
> FROM atlassian/bitbucket:8.11.2
> LABEL maintainer="caelumlux <caelumlux@outlook.com>" version="8.11.2"
> 
> ARG AGENT_PATH=/opt/atlassian
> ARG AGENT_FILENAME=atlassian-agent-jar-with-dependencies.jar
> # 将破解包加入容器
> COPY ./atlassian-agent-jar-with-dependencies.jar /opt/atlassian/
> ENV ATLASSIAN_DIR="/opt/atlassian/bitbucket/app/WEB-INF/lib"
> ENV JAVA_OPTS="-javaagent:${AGENT_PATH}/${AGENT_FILENAME} ${JAVA_OPTS}"
> ```
Appreciate, I will try this as well for testing

voarsh2 commented:
@caelumlux 

> Maybe the agent doesn't work in the DC . You can fix it ....

This is the entire issue..... At 7.21 of Bitbucket, DC licence for INSTALL works, DC licence for plugin works... they've changed something, I am not a Java dev by any means and cannot fix it.

Which is why I opened this pull request, was hoping to make you aware of this.... you made this fork to update it, but it's not working anymore, so needs some more updating. Haha.

Because if they've broken it for Bitbucket, they will probably do the same for the other products (Jira/Confluence/Crowd, etc). Server is ending (which means in 2024 the server licences won't work for new apps, etc.),

caelumlux commented:
@voarsh2 

> Which is why I opened this pull request, was hoping to make you aware of this....

I am not the origin developer of this agent, and I'm not familiar with Atlassian's products either.
So I can't help you either....

voarsh2 commented:
> I am not the origin developer of this agent, and I'm not familiar with Atlassian's products either.
> So I can't help you either....

Well aware, if you aren't familiar with the agent or the products, I am not sure why you made updates to it.

Any case, when server support ends, plugins won't work. DC licences with this agent no longer works past V7.21 of Bitbucket.

Thanks for listening, maybe someone else will pick it up and fix it.

caelumlux commented:
> I am not sure why you made updates to it.

hahahahah, because when I use it myself ,  I found it doen't work before I update ,so I fork it ,and fix some thing simply.
I think maybe someone  will need ,so I push it. 
That's all.

voarsh2 commented:
No worries. I fixed it:
https://github.com/haxqer/bitbucket/issues/1#issuecomment-1623708090.
Will be releasing the changes that fixes DC plugin licencing.

caelumlux commented:
@voarsh2 
oh,that's good,you can push it.

caelumlux commented:
@voarsh2 
oh,sorry,I just find that you have push the code in this pull request now.
I was only notice that Environmet change ,it is not necessary, I'sorry about that.

thank you for your work.

My English is not so well,mabye something goes run.
